### PR TITLE
mod: Bump k8s and update for breaking clustername and serverresources

### PR DESCRIPTION
**Pull request by nfuden**
Minimal k8s 1.25 changes and all dep changes.

### CODE REVIEW

**Comment by Sodman** [comment]
CC @conradhanson who's been working on support for k8s 1.25 for all of the Gloo Platform dependencies

File: changelog/v0.2.0/k8s-1.24+.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+changelog:
+  - type: FIX
+    issueLink: https://github.com/solo-io/solo-projects/issues/7071
```
Comment (jenshu) [inline changelog/v0.2.0/k8s-1.24+.yaml:3]: what is this issue? link doesn't work for me

Comment (jenshu) [reply changelog/v0.2.0/k8s-1.24+.yaml:3]: ah should be gloo repo

File: go.mod [hunk 2]
```diff
@@ -1,56 +1,189 @@
 module github.com/solo-io/k8s-utils
 
-go 1.14
+go 1.19
```
Comment (jenshu) [inline go.mod:3]: do we need to go to 1.19? do we plan to bump our other repos to 1.19?

File: go.mod [hunk 3]
```diff
@@ -1,56 +1,189 @@
...
+	oras.land/oras-go v1.1.0 // indirect
+	sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2 // indirect
+	sigs.k8s.io/kustomize/api v0.11.4 // indirect
+	sigs.k8s.io/kustomize/kyaml v0.13.6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 )
 
 replace (
-	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.0.0+incompatible
+	// github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.0.0+incompatible
```
Comment (jenshu) [inline go.mod:186]: can you confirm why we had this pin before and why we no longer need it?

File: changelog/v0.2.0/k8s-1.24+.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+changelog:
+  - type: FIX
```
Comment (jenshu) [inline changelog/v0.2.0/k8s-1.24+.yaml:2]: somewhere in this file we should include a BREAKING_CHANGE

File: kubeutils/annotations.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package kubeutils
+
+// SoloClusterAnnotation was originally implemented in solo-kitv2's ezkube package
+// It allows for easy mimicking of the old clustername field present in
+// k8s prior to 1.24. It is stored here to make future changes easier and not
+// require multi package switches as if any of the dependant packages are out of
+// sync on this field it can cause hard to diagnose issues.
+const SoloClusterAnnotation = "cluster.solo.io/cluster"
+
+type hasAnnotations interface {
```
Comment (jenshu) [inline kubeutils/annotations.go:10]: nit: interface naming - hasAnnotation sounds more like a function. what do you think of something like ObjectWithAnnotations? also do we need to export it for other repos to use?

File: kubeutils/annotations.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+// SetClusterName on the retrieved annotations
+// Set annotations which while slow is correct.
+func SetClusterName(ha hasAnnotations, clusterName string) {
+	anno := ha.GetAnnotations()
+	if anno == nil {
+		anno = map[string]string{}
+	}
+	anno[SoloClusterAnnotation] = clusterName
+	ha.SetAnnotations(anno)
```
Comment (jenshu) [inline kubeutils/annotations.go:28]: to reduce code duplication, once this PR is merged, can we have skv2 call these functions?

File: kubeutils/annotations.go (hunk 5 above)
Comment (nfuden) [reply kubeutils/annotations.go:10]: This interface is used purely to not have to care about the outside interfaces so id rather not export it so it is implicitly local

File: kubeutils/annotations.go (hunk 6 above)
Comment (nfuden) [reply kubeutils/annotations.go:28]: Yeah if others are ok with that I would prefer that @conradhanson what do you think

File: go.mod (hunk 3 above)
Comment (nfuden) [reply go.mod:186]: I cant see why we used to pin it. Basically this dep was too old and the git blame doesnt state why

File: go.mod (hunk 2 above)
Comment (nfuden) [reply go.mod:3]: Yeah we need 1.19 because k8s mod has a 1.19 reference :( 

This does imply we need the 1.19 bump everywhere

File: go.mod (hunk 3 above)
Comment (nfuden) [reply go.mod:186]: https://github.com/solo-io/go-utils/commit/2af7be12c6cfd50ed5890f59fbf03b4b404c5707#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6
Alright found it again. Autorest used to have an ambiguous import but no longer does

File: kubeutils/annotations.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+// sync on this field it can cause hard to diagnose issues.
+const SoloClusterAnnotation = "cluster.solo.io/cluster"
+
+type annotationStore interface {
+	SetAnnotations(map[string]string)
+	GetAnnotations() map[string]string
+}
+
+// GetClusterName from within the annotations
+func GetClusterName(ha annotationStore) string {
```
Comment (jenshu) [inline kubeutils/annotations.go:16]: nit: change the `ha` variable now that the interface is renamed

File: kubeutils/annotations.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
+package kubeutils
+
+// SoloClusterAnnotation was originally implemented in solo-kitv2's ezkube package
```
Comment (jenshu) [inline kubeutils/annotations.go:3]:
```suggestion
// SoloClusterAnnotation was originally implemented in skv2's ezkube package
```
(use actual repo name)

File: kubeutils/annotations.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+// It allows for easy mimicking of the old clustername field present in
+// k8s prior to 1.24. It is stored here to make future changes easier and not
+// require multi package switches as if any of the dependant packages are out of
```
Comment (jenshu) [inline kubeutils/annotations.go:6]:
```suggestion
// require multi package switches as if any of the dependent packages are out of
```

File: changelog/v0.2.0/k8s-1.24+.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+changelog:
+  - type: BREAKING_CHANGE
+    issueLink: https://github.com/solo-io/gloo/issues/7071
+    resolvesIssue: false
+    description: Clustername override
```
Comment (jenshu) [inline changelog/v0.2.0/k8s-1.24+.yaml:5]:
```suggestion
    description: Use annotations to store/retrieve cluster names, to support kubernetes versions 1.24 and later where the `metadata.clusterName` field does not exist.
```
not sure if this changelog show up anywhere but let's put a more detailed comment

**Review by jenshu — APPROVED** [review]

File: kubeutils/annotations.go (hunk 6 above)
Comment (conradhanson) [reply kubeutils/annotations.go]: Makes sense to me. 
Having those helper functions under the ezkube package of skv2 doesn't exactly scream clarity to me.

**Pull request merged**